django/thunderstore/frontend/api/experimental/tests/test_community_package_list.py

```diff
@@ -0,0 +1,418 @@
+        community_=site.community, package_version_kwargs={"downloads": 23}
+    )
+    listing3 = PackageListingFactory(
+        community_=site.community, package_version_kwargs={"downloads": 42}
+    )
+
+    data = __query_api(
+        api_client, site.community.identifier, "ordering=most-downloaded"
+    )
+
+    __assert_packages_by_listings(data, [listing3, listing2, listing1])
+
+    # Downloads of all versions are counted towards package's downloads.
+    PackageVersionFactory(
+        package=listing1.package, downloads=9001, version_number="1.0.1"
+    )
+    bust_cache_by_condition(CacheBustCondition.any_package_updated)
+
+    data = __query_api(
+        api_client, site.community.identifier, "ordering=most-downloaded"
+    )
+
+    __assert_packages_by_listings(data, [listing1, listing3, listing2])
+
+
+@pytest.mark.django_db
+def test_packages_can_be_ordered_by_rating(api_client: APIClient) -> None:
+    site = CommunitySiteFactory()
+    listing1 = PackageListingFactory(community_=site.community)
+    listing2 = PackageListingFactory(community_=site.community)
+    listing3 = PackageListingFactory(community_=site.community)
+    PackageRatingFactory(package=listing1.package)
+    PackageRatingFactory(package=listing3.package)
+    PackageRatingFactory(package=listing3.package)
+
+    data = __query_api(api_client, site.community.identifier, "ordering=top-rated")
+
+    __assert_packages_by_listings(data, [listing3, listing1, listing2])
+
+
+@pytest.mark.django_db
+def test_pinned_packages_are_returned_first(api_client: APIClient) -> None:
+    site = CommunitySiteFactory()
+    listing1 = PackageListingFactory(community_=site.community)
+    listing2 = PackageListingFactory(
+        community_=site.community, package_kwargs={"is_pinned": True}
+    )
+    listing3 = PackageListingFactory(community_=site.community)
+
+    data = __query_api(api_client, site.community.identifier)
+
+    __assert_packages_by_listings(data, [listing2, listing3, listing1])
+
+
+@pytest.mark.django_db
+def test_deprecated_packages_are_returned_last(api_client: APIClient) -> None:
+    site = CommunitySiteFactory()
+    listing1 = PackageListingFactory(community_=site.community)
+    listing2 = PackageListingFactory(
+        community_=site.community, package_kwargs={"is_deprecated": True}
+    )
+    listing3 = PackageListingFactory(community_=site.community)
+
+    data = __query_api(api_client, site.community.identifier, "deprecated")
+
+    __assert_packages_by_listings(data, [listing3, listing1, listing2])
+
+
+@pytest.mark.django_db
+def test_pagination(api_client: APIClient) -> None:
+    site = CommunitySiteFactory()
+
+    for i in range(25):
+        PackageListingFactory(community_=site.community)
+
+    data = __query_api(api_client, site.community.identifier)
+
+    assert len(data["packages"]) == 24
+
+    data = __query_api(api_client, site.community.identifier, "page=2")
+
+    assert len(data["packages"]) == 1
```

Comment:
Perhaps could use a check making sure the package on page 2 doesn't also get listed on page 1? I've seen some weird pagination errors with APIs where multiple pages return same data (not on this project), so it feels like a sane check to have.

Reply:
Augmenting the test case.